### PR TITLE
[SPARK-4508] [SQL] build native date type to conform behavior to Hive

### PR DESCRIPTION
**Pull request by adrian-wang**
The previous #3732 is reverted due to some test failure.
Have fixed that.

### CODE REVIEW

**Comment by marmbrus** [comment]
Thanks for the quick fix.

LGTM pending tests.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DateUtils.scala [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+import java.sql.Date
+import java.util.{Calendar, TimeZone}
+
+import org.apache.spark.sql.catalyst.expressions.Cast
+
+/**
+ * helper function to convert between Int value of days since 1970-01-01 and java.sql.Date
+ */
+object DateUtils {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DateUtils.scala]: Should we mark it `private[spark]`?

Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DateUtils.scala]: I think this could be useful even outside of spark.

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DateUtils.scala]: If it's not necessary, just do not make it public, or we can not change it anymore.

**Comment by adrian-wang** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DateUtils.scala (hunk 1 above)
Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DateUtils.scala]: Code gen need that.

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DateUtils.scala]: That make sense, thanks!

**Pull request closed without merging**